instrumentation/jdbc/library/src/test/java/io/opentelemetry/instrumentation/jdbc/internal/JdbcConnectionUrlParserTest.java

```diff
@@ -0,0 +1,1237 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.jdbc.internal;
+
+import static io.opentelemetry.instrumentation.jdbc.internal.JdbcConnectionUrlParser.parse;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+
+import io.opentelemetry.instrumentation.jdbc.internal.dbinfo.DbInfo;
+import java.util.Properties;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.ArgumentsProvider;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class JdbcConnectionUrlParserTest {
+  private static Properties stdProps() {
+    Properties prop = new Properties();
+    // https://download.oracle.com/otn-pub/jcp/jdbc-4_1-mrel-spec/jdbc4.1-fr-spec.pdf
+    prop.setProperty("databaseName", "stdDatabaseName");
+    prop.setProperty("dataSourceName", "stdDatasourceName");
+    prop.setProperty("description", "Some description");
+    prop.setProperty("networkProtocol", "stdProto");
+    prop.setProperty("password", "PASSWORD!");
+    prop.setProperty("portNumber", "9999");
+    prop.setProperty("roleName", "stdRoleName");
+    prop.setProperty("serverName", "stdServerName");
+    prop.setProperty("user", "stdUserName");
+    return prop;
+  }
+
+  @ParameterizedTest
+  @ValueSource(strings = {"", "jdbc:", "jdbc::", "bogus:string"})
+  void testInvalidUrlReturnsDefault(String url) {
+    assertThat(JdbcConnectionUrlParser.parse(url, null)).isEqualTo(DbInfo.DEFAULT);
+  }
+
+  @Test
+  void testNullUrlReturnsDefault() {
+    assertThat(JdbcConnectionUrlParser.parse(null, null)).isEqualTo(DbInfo.DEFAULT);
+  }
+
+  @ParameterizedTest(name = "{index}: {2}:{3} parsing of {0}")
+  @ArgumentsSource(ParsingProvider.class)
+  void testVerifySystemSubtypeParsingOfUrl(
```

Comment:
I think this is fine.
Alternatively could try
```java
  @ParameterizedTest(name = "{index}: {0}")
  @ArgumentsSource(ParsingProvider.class)
  void testVerifySystemSubtypeParsingOfUrl(ParseTestArgument argument) {
    DbInfo info = parse(argument.url, argument.properties);
    DbInfo expected = argument.dbInfo;
    assertThat(info.getSystem()).isEqualTo(expected.getSystem());
    assertThat(info.getHost()).isEqualTo(expected.getHost());
    assertThat(info.getPort()).isEqualTo(expected.getPort());
    assertThat(info.getUser()).isEqualTo(expected.getUser());
    assertThat(info.getName()).isEqualTo(expected.getName());
    assertThat(info.getDb()).isEqualTo(expected.getDb());
    assertThat(info).isEqualTo(expected);
  }

  static class ParseTestArgument {
    final String url;
    final Properties properties;
    final DbInfo dbInfo;

    ParseTestArgument(ParseTestArgumentBuilder builder) {
      this.url = builder.url;
      this.properties = builder.properties;
      this.dbInfo =
          DbInfo.builder()
              .system(builder.system)
              .subtype(builder.subtype)
              .user(builder.user)
              .name(builder.name)
              .db(builder.db)
              .host(builder.host)
              .port(builder.port)
              .build();
    }

    @Override
    public String toString() {
      return dbInfo.getSystem() + ":" + dbInfo.getSubtype() + " parsing of " + url;
    }
  }

  static class ParseTestArgumentBuilder {
    String url;
    Properties properties;
    String system;
    String subtype;
    String user;
    String host;
    Integer port;
    String name;
    String db;

    ParseTestArgumentBuilder(String url) {
      this.url = url;
    }

    ParseTestArgumentBuilder setProperties(Properties properties) {
      this.properties = properties;
      return this;
    }

    ParseTestArgumentBuilder setSystem(String system) {
      this.system = system;
      return this;
    }

    ParseTestArgumentBuilder setSubtype(String subtype) {
      this.subtype = subtype;
      return this;
    }

    ParseTestArgumentBuilder setUser(String user) {
      this.user = user;
      return this;
    }

    ParseTestArgumentBuilder setHost(String host) {
      this.host = host;
      return this;
    }

    ParseTestArgumentBuilder setPort(Integer port) {
      this.port = port;
      return this;
    }

    ParseTestArgumentBuilder setName(String name) {
      this.name = name;
      return this;
    }

    ParseTestArgumentBuilder setDb(String db) {
      this.db = db;
      return this;
    }

    ParseTestArgument build() {
      return new ParseTestArgument(this);
    }
  }

  private static ParseTestArgumentBuilder arg(String url) {
    return new ParseTestArgumentBuilder(url);
  }

  static Stream<Arguments> args(ParseTestArgument... testArguments) {
    List<Arguments> list = new ArrayList<>();
    for (ParseTestArgument arg : testArguments) {
      list.add(Arguments.arguments(arg));
    }
    return list.stream();
  }

  static final class ParsingProvider implements ArgumentsProvider {

    @Override
    public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
      return args(
          arg("jdbc:postgresql:///")
              .setSystem("postgresql")
              .setHost("localhost")
              .setPort(5432)
              .build(),
          arg("jdbc:postgresql:///")
              .setProperties(stdProps())
              .setSystem("postgresql")
              .setUser("stdUserName")
              .setHost("stdServerName")
              .setPort(9999)
              .setDb("stdDatabaseName")
              .build());
    }
  }
```

Reply:
this is much easier to read, i'll move things into this format, thanks for the suggestion!